[WIP] Obstacle avoidance docs

hamishwillee opened the pull request.

mrivi commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,120 @@
+
+The flight controller must be in Offboard mode.
+
+The *local_planner* avoidance software must be running on the companion computer.
```
the other algorithm on PX4/avoidance, global_planner, can be used as well

mrivi commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,120 @@
+
+This topic explains how the feature is set up and enabled in both modes.
+
+> **Tip** Manual modes instead use the much simpler *Collision Avoidance* approach, which simply stops the vehicle from crashing into obstacles.
```
This is partially true. As you have written, we have enabled this interface only in AUTO modes. However since the `vehicle_trajectory_waypoint_desired` contains the desired velocity setpoints (generated by any FlightTask), it can be use for full obstacle avoidance in manual mode. I have tested the interface with such a system. However the user would need to modify [this](https://github.com/PX4/Firmware/blob/master/src/modules/mc_pos_control/mc_pos_control_main.cpp#L1187-L1188) and the rate at which the messages are sent from the autopilot. Plus it would need to use an obstacle avoidance algorithm that accepts as inputs velocity setpoints (both local and global planner don't support it). 
I am not sure if it's worth mentioning since it isn't fully supported.

hamishwillee replied:
The main reason I added this at all was because collision avoidance and obstacle avoidance are so similar in name I felt we needed to differentiate. I think I will remove this though, and have the differentiation in a higher level "computer vision" page. 

I wouldn't mention this possibility in either case unless it was actually enabled. In addition, while it might be cool, I suspect that for manual control it would have little real additional value.

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,156 @@
+> **Note** The only required PX4-side setup is to put PX4 into *Offboard mode*..
+  While the `SET_POSITION_TARGET_LOCAL_NED` setpoints come from a ROS collision avoidance node, to PX4 could be from any MAVLink system.
+
+The tested hardware/software platform is [Intel Aero](../flight_controller/intel_aero.md) running either the *local_planner* or *global_planner*. 
```
OK, so my assumption here is that the instructions in PX4/avoidance are good enough for arbitrary person X to find out what hardware they need for avoidance, and how to set up the software for use with this system. I have my doubts :-)

I think it makes sense to offload that to the avoidance docs, because none of that is PX4, and lots of things might use that software and setup.

hamishwillee replied:
What specific configuration do we need to do to get trajectory messages emitted?

mrivi replied:
Do you mean hardware setup?

hamishwillee replied:
@mrivi No, I mean software. For example, to get local planner to emit the right message for collision avoidance we have to use the *rqt_reconfigure* tool and set the parameter `send_obstacles_fcu` to true. 

That said, IMO this feature will remain completely unusable until some work is done on the PX4/avoidance page to better explain how the hardware is set up.

mrivi replied:
it's default

hamishwillee replied:
Thanks. Let's make this clear in global_planner docs.

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,156 @@
+- A waypoint is "reached" when the vehicle is within the acceptance radius, regardless of its heading.
+  - In normal missions the vehicle must reach a waypoint with a certain heading (i.e. in a "close to" straight line from the previous waypoint). 
+  - When obstacle avoidance is active, this constraint cannot be fulfilled because the obstacle avoidance algorithm has full control of the vehicle heading, and the vehicle always moves in the current field of view. 
+- navigator updates the triplets when the vehicle has reached the acceptance radius of each waypoint. 
```
What triplets?

mrivi replied:
navigator works with triplets (previous, current, next) waypoint. The current waypoint is the avoidance goal.

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,156 @@
+PX4 communicates with the computer vision system on the companion computer using an implementation of the [Path Planning Protocol](TBD) (Trajectory Interface). <!-- check we call it this! -->
+
+PX4 sends the desired trajectory to the companion computer in `TRAJECTORY_REPRESENTATION_WAYPOINTS` messages at 5Hz.
+The "waypoint" array fields are set as shown:
```
@mrivi is the uOrb topic for the desired trajectory a 1:1 mapping to the TRAJECTORY_REPRESENTATION_WAYPOINTS? (this is just your text for the uOrb copied as though it was also true for the TRAJECTORY_REPRESENTATION_WAYPOINTS)

I wanted to talk about the interface separate from the implementation.

mrivi replied:
No, they aren't. Here I have described [this array](https://github.com/PX4/Firmware/blob/e91db7b4d2694cd7a1c11b07e979fbbbb5561e99/msg/vehicle_trajectory_waypoint.msg#L19)

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,156 @@
+
+The messages should be sent over the whole mission (not just when navigating around an obstacle).
+The rate at which target setpoints are sent depends on the capabilities of the planning software. 
+Nominally this should exceed [TBD].
```
1. What is the rate at which the local and global planners emit setpoints?
2. What speed does this limit missions to? What happens if a mission has greater speed specified than the obstacle avoidance can meet?
3. What if the vehicle reaches a setpoint before another arrives (what will it do). What if the companion stops sending setpoints altogether?

mrivi replied:
1. local planner ~30Hz, global planner ~10Hz
2. Right now the local planner has a max speed limit of 3m/s. The mission speed set in QGC is not taken into account at the moment. I think when we last tested the global planner outside the speed was around 1-1.5m/s (it was more than a year ago so I am not 100% positive)
3. On the Firmware side there is a check on the setpoints timestamp. If they are too old they're aren't expected anymore.

hamishwillee replied:
Thanks very much. More clarification:
>> What if the vehicle reaches a setpoint before another arrives (what will it do). What if the companion stops sending setpoints altogether?

> On the Firmware side there is a check on the setpoints timestamp. If they are too old they're aren't expected anymore.

OK, but what does the vehicle do? Sit there? Change to Loiter mode? Disable object avoidance and continue the mission. Crash?

mrivi replied:
It disables the obstacle avoidance and the mission continues. (This is going to change shortly)

hamishwillee replied:
Awesome. How long is the delay without messages before mission will continue?

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,156 @@
+
+The paragraphs below describe the behaviour in greater detail, covering the internal PX4 behaviour and message flow through ROS.
+
+#### Mission Mode Detailed Behaviour
```
FYI, this just copies the text you provided. I think it will help people who want to understand the parts of the story.

mrivi replied:
Sure :)

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,164 @@
+
+PX4 sends the desired trajectory to the companion computer in [TRAJECTORY_REPRESENTATION_WAYPOINTS](https://mavlink.io/en/messages/common.html#TRAJECTORY_REPRESENTATION_WAYPOINTS) messages at 5Hz.
+
+The fields are set as shown:
```
FYI, updated the interface to refer to the actual message fields, not the uORB fields. Please check.

mrivi replied:
looks correct

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,164 @@
+PX4 sends the desired trajectory to the companion computer in [TRAJECTORY_REPRESENTATION_WAYPOINTS](https://mavlink.io/en/messages/common.html#TRAJECTORY_REPRESENTATION_WAYPOINTS) messages at 5Hz.
+
+The fields are set as shown:
+- `time_usec`: UNIX Epoch time.
```
Is UNIX Epoch time correct?

hamishwillee commented:
Thanks @mrivi This has had another round of update.
A couple of questions left:
- https://github.com/PX4/px4_user_guide/pull/376/files#r232524313
- https://github.com/PX4/px4_user_guide/pull/376#discussion_r232527554

I don't think this document is far off being ready to publish (what do you think?)

However to be useful I think a LOT more information is needed on hardware setup. This should be in PX4/avoidance repo. Thoughts on this? I can't really do much on this because I don't have hardware to test it. @baumanta Said she might take a shot for on this for the collision avoidance stuff, but would be great if you could collaborate.
[A newbie should be able to find all the equipment and software needed pretty easily and set this up in a "turnkey" way]

hamishwillee commented:
@mrivi Ping :-)

hamishwillee commented on en/computer_vision/obstacle_avoidance.md:
```diff
@@ -0,0 +1,166 @@
+  - *local planner* mission speed is around 3 m/s.
+  - *global planner* mission speed is around 1-1.5 m/s.
+
+If PX4 stops receiving setpoint updates then obstacle avoidance will be disabled, and the mission continues under normal PX4 [Mission mode](../flight_modes/mission.md) control. 
```
@mrivi FYI, this is where we would add "how long after obstacle avoidance messages stops that we got back to normal mission".

How does the system notify the user of this problem?

mrivi replied:
0.5seconds. Right now there is no notification. Tanja is working on that right now.

hamishwillee commented:
@mrivi Firstly, thanks for the review - I appreciate you are busy and that docs are a pain.

This is good enough to merge, so I did so - Check it out [here](https://docs.px4.io/en/computer_vision/obstacle_avoidance.html)!

But, the following details still requiring an answer:
- https://github.com/PX4/px4_user_guide/pull/376#discussion_r233695522
- https://github.com/PX4/px4_user_guide/pull/376#discussion_r232531114

The one thing that could improve this IMO is a diagram, at least for the mission case (but ideally both). This would show:
- outer boxes for FCU, Companion, Camera
- inner boxes for PX4, local planner/global planner, camera nodes, any other ROS nodes
- message markup for ingoing /outgoing.
Do you have something like that or would you like to try create it? I can do first draft if not.

mrivi commented:
@hamishwillee makes sense to me. If you have the time to start creating the diagrams it would be super helpful. 
The global planner can be used by any system that talks mavlink and sends/receives the messages the planner expects. The local planner is slightly more coupled with PX4 since it has knowledge on the vehicle mode. I am not familiar with Ardupilot so I don't know if the modes such as auto mission, offboard, auto land are the same. 

@baumanta have you already started wring up something? If so can you share it as WIP ?

hamishwillee commented:
@mrivi Great. Then I'll leave you and @baumanta to talk about those docs until you need my further help.

Re a diagram, here is a starting point. You're the expert, what is missing :-) ?

![image](https://user-images.githubusercontent.com/5368500/48680619-fa5edf00-ebf0-11e8-9110-3fad17dd0e40.png)

baumanta commented:
@hamishwillee all those topics also go from PX4 to the planner:

 /mavros/local_position/pose
/mavros/local_position/velocity
/mavros/state
/mavros/trajectory/desired
/mavros/altitude (optional)
/tf

hamishwillee commented:
@baumanta Unless you've got this working with ROS2, PX4 only sends mavlink messages, not topics. Do you know which message those correspond to?

We could add nodes between avoidance and show the intermediate topics, but you'd need to make it clear what they are.

mrivi commented:
![image from ios 1](https://user-images.githubusercontent.com/15078736/48758615-ee325a80-eca0-11e8-8b86-9b2a3c26bb61.jpg)

@hamishwillee  This is the complete flow from Firmware to local planner.

mrivi commented:
![image from ios 2](https://user-images.githubusercontent.com/15078736/48759559-c395d100-eca3-11e8-9911-2e64e2ed800e.jpg)

From local_planner to Firmware. The global planner only sends a position setpoint in offboard to the Firmware

hamishwillee commented:
@mrivi Thanks very much. Without knowing very much about ROS this sort of message flow information seems pretty useful. I'll try to put this in a diagram, but in the meantime, we could minimally have as a table - maybe best in the avoidance docs?

A few questions/confirmation. 
- What is the PX4 uorb topic under home_position (vehicle_air???)
- When you say "This is the complete flow from PX4 to the `local_planner`." 
  - I presume you mean that the final "ROS topic" is what the local planner uses? ie so in a diagram you'd have a line from the ROS topic to something called "local_planner" (which is a "ROS Plugin?)
- "The global planner only needs LOCAL_POSITION_NED" 
   -Are you saying that the input to the global planner looks like this:

PX4 topic | MAVLink | MAVROS Plugin | ROS Msgs. | Topic 
--- | --- | --- | --- | ---
vehicle_local_position | LOCAL_POSITION_NED | local_position | geometry_msgs::PoseStomped | mavros/local_position/pose

* "The global planner only sends a position setpoint in offboard to the Firmware"
  - Does that mean the whole table for offboard to firmware would be:

ROS topic | ROS Msgs. | MAVROS Plugin | MAVLink | PX4 Topic 
--- | --- | --- | --- | ---
/mavros/setpoint_position/local (offboard) | geometry_msgs::PoseStomped | setpoint_position | SET_POSITION_LOCAL_POSITION_NED | position_setpoint_triplet


Tables below captured to remove any spelling or other issues. Let me know if correct please.

---

This is the complete flow from PX4 to the `local_planner`. The global planner only needs LOCAL_POSITION_NED

PX4 topic | MAVLink | MAVROS Plugin | ROS Msgs. | ROS Topic 
--- | --- | --- | --- | ---
vehicle_local_position | LOCAL_POSITION_NED | local_position | geometry_msgs::PoseStomped | mavros/local_position/pose
vehicle_local_position | LOCAL_POSITION_NED | local_position | geometry_msgs::TwistStomped | mavros/local_position/velocity
vehicle_local_position | ALTITUDE | altitude | mavros_msgs::Altitude | mavros/altitude 
home_position | ALTITUDE | altitude | mavros_msgs::Altitude | mavros/altitude
vehicle_air??? | ALTITUDE | altitude | mavros_msgs::Altitude | mavros/altitude
vehicle_status | HEARTBEAT | sys_status | mavros_msgs::State | mavros/state
vehicle_trajectory_waypoint_desired | TRAJECTORY_REPRESENTATION_WAYPOINT | trajectory  | mavros_msgs::Trajectory | mavros/trajectory/desired

From local_planner to PX4. The global planner only sends a position setpoint in offboard to the Firmware

ROS topic | ROS Msgs. | MAVROS Plugin | MAVLink | PX4 Topic 
--- | --- | --- | --- | ---
/mavros/setpoint_position/local (offboard) | geometry_msgs::PoseStomped | setpoint_position | SET_POSITION_LOCAL_POSITION_NED | position_setpoint_triplet
/mavros/setpoint_velocity/cmd_vel_unstamped (offboard) | geometry_msgs::TwistStomped | setpoint_velocity | SET_POSITION_LOCAL_POSITION_NED | position_setpoint_triplet
/mavros/trajectory/generated (mission) | mavros_msgs::Trajectory | trajectory | TRAJECTORY_REPRESENTATION_WAYPOINT | vehicle_trajectory_waypoint
/mavros/obstacle/send | sensor_msgs::LaserScan | obstacle_distance | OBSTACLE_DISTANCE | obstacle_distance
/mavros/set_mode | mavros_msgs::SetMode | sys_status | SET_MODE | vehicle_command

mrivi commented:
- It's `vehicle_air_data`
- yes, the local planner (or global planner) are ROS nodes are they subscribe/publish to the ROS topics listed above
- in the first table `mavros/local_position/velocity` is missing

hamishwillee commented:
Thanks @mrivi. In my comment I fixed up `vehicle_air_data` and added mavros/local_position/velocity to the global planner first table (please check). 

The most important question is probably the first one:
- Do you see any real benefit in trying to put this in a diagram like the one I created above? 
- I like this table - do you think it should live in the avoidance docs, PX4 docs or both? My leaning this table in  avoidance docs, with a simplified diagram in the PX4 docs.

mrivi commented:
@hamishwillee thank you. I realized there is a typo `geometry_msgs::PositionStamped` not `geometry_msgs::PositionStomped`. Same for the Twist message.

Table in the avoidance docs and simpler diagram in the PX4 docs sounds good to me

baumanta commented:
@hamishwillee I really like those tables! Would have made my life easier a few months ago ;)

hamishwillee commented:
Awesome! I've added tables to PX4/avoidance readme: https://github.com/PX4/avoidance/pull/145 . Feel free to modify or move as you need to (ie as discussed, this readme could be split up). 
I've added an issue to possibly add a diagramatic representation in the user guide in https://github.com/PX4/px4_user_guide/issues/386


> I really like those tables! Would have made my life easier a few months ago ;)

@baumanta Me too - thanks very much @mrivi for the tables. Hopefully they will remain generally useful going forward!

mrivi commented:
@hamishwillee I am working on a refactor of the avoidance readme. The work in progress PR will be up shortly

hamishwillee commented:
Thanks for letting me know @mrivi - I can see today that you've made some changes. Will try check them out today or tomorrow.
